File: luaSource/unit_test/test_forward_argument.lua
--测试lua接收到的oc传入的参数


local crash = function (message)
	-- body
	print("program crash")
	print(message)
	os.exit(1)
end
--这里的assert好像不能检测错误
_G.TestForwardArgument_checkInt = function (className,i)
	-- body
	-- assert(i == 1,"TestForwardArgument_checkInt error")
	if i ~= 1 then
		crash("TestForwardArgument_checkInt error")
	end
	return i
end

_G.TestForwardArgument_checkDouble = function (className,i)
	-- body
	-- assert(i == 1.1,"TestForwardArgument_checkDouble error")
	if i ~= 1.1 then
		crash("TestForwardArgument_checkDouble error")
	end
	return i
end
_G.TestForwardArgument_checkId = function (className,i)
	-- body
	-- assert(i.isKindOfClass(require("luapatch").packClass("NSDictionary").class()) == 1,"TestForwardArgument_checkId error")
	if i.isKindOfClass(require("luapatch").packClass("NSDictionary").class()) ~= 1 then
		crash("TestForwardArgument_checkId error")
	end
	return i
end
_G.TestForwardArgument_checkSEL = function (className,i)
	-- body
	-- assert(i == "checkSEL:","TestForwardArgument_checkSEL error")
	print("call TestForwardArgument_checkSEL in lua")
	if i ~= "checkSEL:" then
		crash("TestForwardArgument_checkSEL error")
	end
	return i
end
_G.TestForwardArgument_checkPoint = function (className,i)
	-- body
	-- assert(require("luapatch").convertUserDataToStr(i) == "hello world","TestForwardArgument_checkPoint error")
	if require("luapatch").convertUserDataToStr(i) ~= "hello world" then
		crash("TestForwardArgument_checkPoint error")
	end
	return i
end
_G.TestForwardArgument_checkClass = function (className,i)
	-- body
	local luapatch = require("luapatch")
	-- assert(luapatch.unpackPoint(i) == luapatch.unpackPoint(luapatch.packClass("NSDictionary").class()),"TestForwardArgument_checkClass error")
	if luapatch.unpackPoint(i) ~= luapatch.unpackPoint(luapatch.packClass("NSDictionary").class()) then
		crash("TestForwardArgument_checkClass error")
	end
	return i
end
_G.TestForwardArgument_checkNum = function (className,i)
	-- body
	-- assert(i == 1,"TestForwardArgument_checkNum error")
	if i ~= 1 then
		crash("TestForwardArgument_checkNum error")
	end
	return i
end
_G.TestForwardArgument_checkStr = function (className,i)
	-- body
	-- assert(i == "hello world","TestForwardArgument_checkStr error")
	if i ~= "hello world" then
		crash("TestForwardArgument_checkStr error")
	end
	return i
end


-- expose("expose lua function")

describe("test forward argument",function( ... )
	-- body
	local luapatch = require("luapatch")
	it("start test forward argument",function ( ... )
		-- body
		luapatch.redefineClassMethod("TestForwardArgument","checkInt:","TestForwardArgument_checkInt")
		luapatch.redefineClassMethod("TestForwardArgument","checkDouble:","TestForwardArgument_checkDouble")
		luapatch.redefineClassMethod("TestForwardArgument","checkId:","TestForwardArgument_checkId")
		luapatch.redefineClassMethod("TestForwardArgument","checkSEL:","TestForwardArgument_checkSEL")
		luapatch.redefineClassMethod("TestForwardArgument","checkPoint:","TestForwardArgument_checkPoint")
		luapatch.redefineClassMethod("TestForwardArgument","checkClass:","TestForwardArgument_checkClass")
		luapatch.redefineClassMethod("TestForwardArgument","checkNum:","TestForwardArgument_checkNum")
		luapatch.redefineClassMethod("TestForwardArgument","checkStr:","TestForwardArgument_checkStr")
		local TestForwardArgument = luapatch.packClass("TestForwardArgument")
		TestForwardArgument.startCheckArgument()
	end)
end)

File: luaPatchDemo/TestForwardArgument.m
//
//  TestForwardArgument.m
//  luaPatchDemo
//
//  Created by 黄钊 on 2019/9/7.
//  Copyright © 2019 hz. All rights reserved.
//

#import "TestForwardArgument.h"

@implementation TestForwardArgument

+(int)checkInt:(int)i{
    return 0;
}

+(double)checkDouble:(double)d{
    return 0;
}

+(id)checkId:(id)i{
    return nil;
}

+(SEL)checkSEL:(SEL)sel{
    return @selector(a:);
}

+(void*)checkPoint:(void *)p{
    return NULL;
}

+(Class)checkClass:(Class)cls{
    return nil;
}

+(NSNumber*)checkNum:(NSNumber*)num{
    return @0;
}

+(NSString*)checkStr:(NSString*)str{
    return nil;
}

+(void)startCheckArgument{
    //测试返回值
    NSAssert([self.class checkInt:1] == 1,@"TestForwardArgument checkInt error");
    NSAssert([self.class checkDouble:1.1] == 1.1,@"TestForwardArgument checkDouble error");
    NSAssert([[self.class checkId:[NSDictionary new]] isKindOfClass:NSDictionary.class],@"TestForwardArgument checkId error");
    NSAssert([NSStringFromSelector([self.class checkSEL:@selector(checkSEL:)]) isEqualToString:@"checkSEL:"],@"TestForwardArgument checkSEL error");
    NSAssert(strcmp([self.class checkPoint:"hello world"], "hello world") == 0,@"TestForwardArgument checkPoint error");
    NSAssert([self.class checkClass:NSDictionary.class] == NSDictionary.class,@"TestForwardArgument checkClass error");
    NSAssert([[self.class checkNum:@1] isEqualToNumber:@1], @"TestForwardArgument checkNum error");
    NSAssert([[self.class checkStr:@"hello world"] isEqualToString:@"hello world"],@"TestForwardArgument checkStr error");    
}

@end
